### PR TITLE
WT-4631 Setting read timestamp to WT_TS_NONE in clear readtimestamp

### PR DESCRIPTION
**Pull request by luke-pearson**
…unction

### CODE REVIEW

**Comment by luke-pearson** [comment]
I think this results in a very low chance race in txn_timestamp.c at line 199, which could possibly be fixed by storing txn->read_timestamp in a temp variable prior to accessing it.

**Comment by michaelcahill** [comment]
@luke-pearson, I think WT-4621 added another similar case here: https://github.com/wiredtiger/wiredtiger/blob/develop/src/txn/txn_timestamp.c#L783-L789

**Comment by michaelcahill** [comment]
Maybe consider something like this:

```static void
__txn_get_read_timestamp(WT_TXN *txn, wt_timestamp_t *read_timestampp, boolean *validp)
{
    *read_timestampp = txn->read_timestamp;
    WT_ORDERED_READ(*validp, !txn->clean_read_q);
}```

File: src/txn/txn_timestamp.c [hunk 1]
```diff
@@ -187,6 +187,8 @@ __txn_get_pinned_timestamp(
 	/* Look for the oldest ordinary reader. */
 	__wt_readlock(session, &txn_global->read_timestamp_rwlock);
 	TAILQ_FOREACH(txn, &txn_global->read_timestamph, read_timestampq) {
+		/* Copy value out to prevent possible race condition. */
+		tmp_read_timestamp = txn->read_timestamp;
```
Comment (agorrod) [inline src/txn/txn_timestamp.c]: The compiler can optimize this assignment away (similarly below). You could either add an explicit read barrier or mark the variable declaration volatile (https://barrgroup.com/Embedded-Systems/How-To/C-Volatile-Keyword).

This is subtle - I'd be happy to talk through the reasons and alternatives tomorrow if you'd like.

Comment (luke-pearson) [reply src/txn/txn_timestamp.c]: I've pushed up a change which uses read barriers, let me know what you think. I'd be interested in having a discussion about them though tomorrow.

**Comment by luke-pearson** [comment]
retest this please

**Pull request merged**